File: 125-valid-palindrome/125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        """
        A man, a plan, a canal: Panama
             i                   j
        A man, a plan, a canal: Panama
        """
        i, j = 0, len(s) - 1
        while i < j:
            l_char, r_char = s[i].lower(), s[j].lower()
            if not l_char.isalnum():
                i += 1
                continue
            if not r_char.isalnum():
                j -= 1
                continue
            if l_char != r_char:
                return False
            i += 1
            j -= 1
        return True
            
            

File: 125-valid-palindrome/125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        String str = s.toLowerCase();
        
        int l = 0;
        int r = str.length() - 1;
        while (l < r) {
            char lChar = str.charAt(l);
            char rChar = str.charAt(r);
            
            if (!Character.isLetterOrDigit(lChar)) {
                l += 1;
            } else if (!Character.isLetterOrDigit(rChar)) {
                r -= 1;
            } else {
                if (lChar != rChar) return false;
                l += 1;
                r -= 1;
            }
        }
        
        return true;
    }
}